File: TA/Frontend/RHStoreWS/RHStoreComprasBO/PayPalAPI.cs
using RHStoreBaseBO.ServiciosWeb;
using System;
using System.Collections.Generic;
using System.Data.SqlTypes;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;

namespace RHStoreComprasBO
{
    public class PayPalAPI
    {
        private string clientId;
        private string clientSecret;
        public PayPalAPI()
        {
            this.clientId = "AT3qKPWFLqjokKJcgBtV7T54TyFzYNtG8HUiC6-ba2Q-r0F2Iyc8LIlaRqjv-9yfOZAckq_WCnfE9A3z";            
            this.clientSecret = "EKrBCfI3mpICyV9cC3Gu7vaNVfBOSspPaJhfcDMJ-W9TC1HaHrCsw48touO0cHVPOc1IzXhe5G4ffsG5";
        }

        public string get_token()
        {
            string token = "-";
            string url = "https://api-m.sandbox.paypal.com/v1/oauth2/token";
            string grantType = "grant_type=client_credentials";
            string credentials = $"{this.clientId}:{this.clientSecret}";
            string encodedCredentials = Convert.ToBase64String(Encoding.UTF8.GetBytes(credentials));

            try
            {
                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
                request.Method = "POST";
                request.ContentType = "application/x-www-form-urlencoded";
                request.Headers["Authorization"] = $"Basic {encodedCredentials}";

                byte[] bodyData = Encoding.UTF8.GetBytes(grantType);
                request.ContentLength = bodyData.Length;
                using (Stream requestStream = request.GetRequestStream())
                {
                    requestStream.Write(bodyData, 0, bodyData.Length);
                }

                using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        string responseText = reader.ReadToEnd();
                        token = this.extract_access_token(responseText);
                    }
                }
            }
            catch (WebException ex)
            {
                using (var errorResponse = (HttpWebResponse)ex.Response)
                {
                    using (var reader = new StreamReader(errorResponse.GetResponseStream()))
                    {
                        string errorText = reader.ReadToEnd();
                        Console.WriteLine("Error:");
                        Console.WriteLine(errorText);
                    }
                }
            }

            return token;
        }

        private string extract_access_token(string json)
        {
            string tokenKey = "\"access_token\":\"";
            int tokenStartIndex = json.IndexOf(tokenKey) + tokenKey.Length;
            if (tokenStartIndex < tokenKey.Length)
            {
                return null;
            }

            int tokenEndIndex = json.IndexOf("\"", tokenStartIndex);
            if (tokenEndIndex == -1) 
            {
                return null;
            }

            return json.Substring(tokenStartIndex, tokenEndIndex - tokenStartIndex);
        }

        public string create_order(int requestId, int referenceId,
            string currencyCode, double value, string returnUrl, string cancelUrl,
            string brandName)
        {
            string link = "-";
            string url = "https://api-m.sandbox.paypal.com/v2/checkout/orders";
            string jsonBody = $@"
            {{
                ""intent"": ""CAPTURE"",
                ""purchase_units"": [
                    {{
                        ""reference_id"": ""{referenceId.ToString()}"",
                        ""amount"": {{
                            ""currency_code"": ""{currencyCode}"",
                            ""value"": ""{value.ToString()}""
                        }}
                    }}
                ],
                ""payment_source"": {{
                    ""paypal"": {{
                        ""experience_context"": {{
                            ""payment_method_preference"": ""IMMEDIATE_PAYMENT_REQUIRED"",
                            ""brand_name"": ""{brandName}"",
                            ""locale"": ""en-US"",
                            ""landing_page"": ""LOGIN"",
                            ""shipping_preference"": ""NO_SHIPPING"",
                            ""user_action"": ""PAY_NOW"",
                            ""return_url"": ""{returnUrl}"",
                            ""cancel_url"": ""{cancelUrl}""
                        }}
                    }}
                }}
            }}";

            try
            {
                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
                request.Method = "POST";
                request.ContentType = "application/json";
                request.Headers["PayPal-Request-Id"] = requestId.ToString();
                request.Headers["Authorization"] = "Bearer " + this.get_token();

                byte[] bodyData = Encoding.UTF8.GetBytes(jsonBody);
                request.ContentLength = bodyData.Length;

                using (Stream requestStream = request.GetRequestStream())
                {
                    requestStream.Write(bodyData, 0, bodyData.Length);
                }

                using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        string responseText = reader.ReadToEnd();
                        link = this.extract_link(responseText);
                    }
                }
            }
            catch (WebException ex)
            {
                using (var errorResponse = (HttpWebResponse)ex.Response)
                {
                    using (var reader = new StreamReader(errorResponse.GetResponseStream()))
                    {
                        string errorText = reader.ReadToEnd();
                        Console.WriteLine("Error:");
                        Console.WriteLine(errorText);
                    }
                }
            }

            return link;
        }

        public string extract_link(string jsonResponse)
        {
            int index_start = jsonResponse.IndexOf("https://www.sandbox.paypal.com");
            string url_dirty = jsonResponse.Substring(index_start, 90);
            int i;
            for ( i = 0; i < url_dirty.Length; i++)
            {
                if (url_dirty[i] == '"')
                {
                    break;
                }
            }
            
            return url_dirty.Substring(0, i);
        }

        public bool capture_order(string order_id)
        {
            bool completed = false;
            //string order_id = this.extract_order_id_from_link(link);
            string responseText = string.Empty;
            string url = $"https://api-m.sandbox.paypal.com/v2/checkout/orders/{order_id}/capture";

            string requestBody = "{}";

            try
            {
                HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
                request.Method = "POST";
                request.ContentType = "application/json";
                request.Headers["Authorization"] = "Bearer " + this.get_token();

                request.Headers["PayPal-Request-Id"] = order_id;

                // Write the request body if necessary
                byte[] bodyData = Encoding.UTF8.GetBytes(requestBody);
                request.ContentLength = bodyData.Length;
                using (Stream requestStream = request.GetRequestStream())
                {
                    requestStream.Write(bodyData, 0, bodyData.Length);
                }

                using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                {
                    using (StreamReader reader = new StreamReader(response.GetResponseStream()))
                    {
                        responseText = reader.ReadToEnd();
                        completed = ("COMPLETED" == this.extract_status(responseText));
                    }
                }
            }
            catch (WebException ex)
            {
                using (var errorResponse = (HttpWebResponse)ex.Response)
                {
                    using (var reader = new StreamReader(errorResponse.GetResponseStream()))
                    {
                        string errorText = reader.ReadToEnd();
                        Console.WriteLine("Error:");
                        Console.WriteLine(errorText);
                    }
                }
            }

            return completed;
        }
        public string extract_order_id_from_link(string url)
        {
            string token = string.Empty;
            try
            {
                int tokenIndex = url.IndexOf("token=");

                if (tokenIndex != -1)
                {
                    tokenIndex += 6;
                    int endIndex = url.IndexOf("&", tokenIndex);

                    if (endIndex == -1)  
                    {
                        endIndex = url.Length;
                    }

                    token = url.Substring(tokenIndex, endIndex - tokenIndex);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error extracting token: {ex.Message}");
            }

            return token;
        }

        public string extract_status(string jsonResponse)
        {
            string statusKey = "\"status\":\"";
            int statusIndex = jsonResponse.IndexOf(statusKey);

            if (statusIndex != -1)
            {
                int startIndex = statusIndex + statusKey.Length;
                int endIndex = jsonResponse.IndexOf("\"", startIndex);

                if (endIndex != -1)
                {
                    return jsonResponse.Substring(startIndex, endIndex - startIndex);
                }
            }

            return string.Empty;
        }
    }
}

File: TA/Backend/SoftRhDBManager/src/pe/edu/pucp/softrh/database/config/PayPalAPI.java
package pe.edu.pucp.softrh.database.config;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.Base64;

public class PayPalAPI {

    private String url_base;
    private String id_cliente;
    private String secreto;
    private String autorizacion;
    private String autorizacion_cifrada;
    private String token_acceso;
    private String orderID;
    private HttpClient cliente;

    private String url_return;
    private String order_id;

    public PayPalAPI( ) {
        this.url_base = "https://api-m.sandbox.paypal.com/";
        this.cliente = HttpClient.newHttpClient();
        this.set_idcliente_secreto("AT3qKPWFLqjokKJcgBtV7T54TyFzYNtG8HUiC6-ba2Q-r0F2Iyc8LIlaRqjv-9yfOZAckq_WCnfE9A3z",
                "EKrBCfI3mpICyV9cC3Gu7vaNVfBOSspPaJhfcDMJ-W9TC1HaHrCsw48touO0cHVPOc1IzXhe5G4ffsG5");
    }

    public void set_idcliente_secreto(String clientId, String secret) {
        this.id_cliente = clientId;
        this.secreto = secret;
        this.autorizacion = this.id_cliente + ":" + this.secreto;
        this.autorizacion_cifrada = Base64.getEncoder().encodeToString(autorizacion.getBytes());
    }

    private ArrayList<String>[] extraer_datos_json(String respuesta) {
        respuesta = respuesta.trim().substring(1, respuesta.length() - 1);
        String[] pairs = respuesta.split(",");

        ArrayList<String> keys = new ArrayList<>();
        ArrayList<String> values = new ArrayList<>();

        for (String pair : pairs) {
            String[] keyValue = pair.split(":", 2);

            String key = keyValue[0].trim().replace("\"", "");
            String value = keyValue[1].trim().replace("\"", "");

            keys.add(key);
            values.add(value);
        }

        return new ArrayList[]{keys, values};
    }

    public ArrayList<String>[] obtener_token_acceso(String url_autorizacion) throws IOException, InterruptedException {
        HttpRequest requerimiento = HttpRequest.newBuilder()
                .uri(URI.create(this.url_base + url_autorizacion))
                .header("Authorization", "Basic " + this.autorizacion_cifrada)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString("grant_type=client_credentials"))
                .build();
        HttpResponse<String> response = this.cliente.send(requerimiento, HttpResponse.BodyHandlers.ofString());

        //System.out.println("Response code: " + response.statusCode());
        ArrayList<String>[] result = this.extraer_datos_json(response.body());
        ArrayList<String> values = result[1];
        this.token_acceso = values.get(1);

        return result;
    }

    public String crear_datos_pago(String senderBatchId,
            String emailSubject, String emailMessage,
            String amountValue, String currency,
            String note, String senderItemId, String receiver) {
        String recipientType = "EMAIL";
        StringBuilder payoutJson = new StringBuilder();

        payoutJson.append("{")
                .append("\"sender_batch_header\": {")
                .append("\"sender_batch_id\": \"").append(senderBatchId).append("\",")
                .append("\"email_subject\": \"").append(emailSubject).append("\",")
                .append("\"email_message\": \"").append(emailMessage).append("\"")
                .append("},")
                .append("\"items\": [")
                .append("{")
                .append("\"recipient_type\": \"").append(recipientType).append("\",")
                .append("\"amount\": {")
                .append("\"value\": \"").append(amountValue).append("\",")
                .append("\"currency\": \"").append(currency).append("\"")
                .append("},")
                .append("\"note\": \"").append(note).append("\",")
                .append("\"sender_item_id\": \"").append(senderItemId).append("\",")
                .append("\"receiver\": \"").append(receiver).append("\"")
                .append("}")
                .append("]}");

        return payoutJson.toString();
    }

    public ArrayList<String>[] realizar_pago(String senderBatchId,
            String emailSubject, String emailMessage,
            String amountValue, String currency,
            String note, String senderItemId, String receiver) throws IOException, InterruptedException {
        String datos_pago = this.crear_datos_pago(senderBatchId,
                emailSubject, emailMessage,
                amountValue, currency,
                note, senderItemId, receiver);
        ArrayList<String>[] respuesta_token = this.obtener_token_acceso("v1/oauth2/token");
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api-m.sandbox.paypal.com/v1/payments/payouts"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + this.token_acceso)
                .POST(HttpRequest.BodyPublishers.ofString(datos_pago))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        //System.out.println("Response code: " + response.statusCode());
        ArrayList<String>[] result = this.extraer_datos_json(response.body());
        ArrayList<String> values = result[1];

        return result;
    }

    public String createOrderJson(String referenceId, String currencyCode, String value,
            String returnUrl, String cancelUrl, String brandName) {

        StringBuilder jsonBuilder = new StringBuilder();

        jsonBuilder.append("{")
                .append("\"intent\": \"CAPTURE\",")
                .append("\"purchase_units\": [")
                .append("{")
                .append("\"reference_id\": \"").append(referenceId).append("\",")
                .append("\"amount\": {")
                .append("\"currency_code\": \"").append(currencyCode).append("\",")
                .append("\"value\": \"").append(value).append("\"")
                .append("}")
                .append("}")
                .append("],")
                .append("\"payment_source\": {")
                .append("\"paypal\": {")
                .append("\"experience_context\": {")
                .append("\"payment_method_preference\": \"IMMEDIATE_PAYMENT_REQUIRED\",")
                .append("\"brand_name\": \"").append(brandName).append("\",")
                .append("\"locale\": \"en-US\",")
                .append("\"landing_page\": \"LOGIN\",")
                .append("\"shipping_preference\": \"NO_SHIPPING\",")
                .append("\"user_action\": \"PAY_NOW\",")
                .append("\"return_url\": \"").append(returnUrl).append("\",")
                .append("\"cancel_url\": \"").append(cancelUrl).append("\"")
                .append("}")
                .append("}")
                .append("}")
                .append("}");

        return jsonBuilder.toString();
    }

    public String sendOrderRequest(String requestId, String referenceId,
            String currencyCode, String value,
            String returnUrl, String cancelUrl, String brandName) throws IOException, InterruptedException {
        String datos = this.createOrderJson(referenceId, currencyCode, value,
                returnUrl, cancelUrl, brandName);
        ArrayList<String>[] respuesta_token = this.obtener_token_acceso("v1/oauth2/token");
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create("https://api-m.sandbox.paypal.com/v2/checkout/orders"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + this.token_acceso)
                .header("PayPal-Request-Id", requestId)
                .POST(HttpRequest.BodyPublishers.ofString(datos))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        //System.out.println("Response code: " + response.statusCode());

        ArrayList<String>[] result = this.extraer_datos_json(response.body());
        ArrayList<String> keys = result[0];
        ArrayList<String> values = result[1];
        this.orderID = values.get(0);

        return values.get(6);
    }

    private String extractStatusFromJson(String json) {

        String statusKey = "\"status\":\"";
        int statusKeyPos = json.indexOf(statusKey);
        if (statusKeyPos != -1) {
            int statusStart = statusKeyPos + statusKey.length();
            int statusEnd = json.indexOf("\"", statusStart);
            if (statusEnd != -1) {
                return json.substring(statusStart, statusEnd);
            }
        }
        return "";
    }

    public Boolean verificate_pay(String url_return) throws URISyntaxException, IOException, InterruptedException {
        String tokenKey = "token=";
        int tokenStartIndex = url_return.indexOf(tokenKey) + tokenKey.length();
        String order_id = url_return.substring(tokenStartIndex);
        ArrayList<String>[] respuesta_token = this.obtener_token_acceso("v1/oauth2/token");
        String url = "https://api-m.sandbox.paypal.com/v2/checkout/orders/" + order_id + "/capture";
        HttpClient client = HttpClient.newHttpClient();

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + this.token_acceso)
                .header("PayPal-Request-Id", order_id)
                .POST(HttpRequest.BodyPublishers.ofString(""))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        String json = response.body();
        String status = this.extractStatusFromJson(json);
        return "COMPLETED".equals(status);
    }
}
